micro_http/src/server.rs

```diff
@@ -0,0 +1,752 @@
+            // `epoll` event set to notify us when the stream is ready for writing.
+            if let ClientConnectionState::AwaitingIncoming = client_connection.state {
+                client_connection.state = ClientConnectionState::AwaitingOutgoing;
+                Self::epoll_mod(self.epoll_fd, response.id as RawFd, epoll::Events::EPOLLOUT)?;
+            }
+            client_connection.enqueue_response(response.response);
+        }
+        Ok(())
+    }
+
+    /// Accepts a new incoming connection and adds it to the `epoll` notification structure.
+    fn handle_new_connection(&mut self) -> Result<()> {
+        if self.connections.len() == MAX_CONNECTIONS {
+            // If we want a replacement policy for connections
+            // this is where we will have it.
+            return Err(ServerError::ServerFull);
+        }
+
+        self.socket
+            .accept()
+            .map_err(ServerError::IOError)
+            .and_then(|(stream, _)| {
+                // `HttpConnection` is supposed to work with non-blocking streams.
+                stream
+                    .set_nonblocking(true)
+                    .map(|_| stream)
+                    .map_err(ServerError::IOError)
+            })
+            .and_then(|stream| {
+                // Add the stream to the `epoll` structure and listen for bytes to be read.
+                Self::epoll_add(self.epoll_fd, stream.as_raw_fd())?;
+                // Then add it to our open connections.
+                self.connections.insert(
+                    stream.as_raw_fd(),
+                    ClientConnection::new(HttpConnection::new(stream)),
+                );
+                Ok(())
+            })
+    }
+
+    /// Changes the event type for a connection to either listen for incoming bytes
+    /// or for when the stream is ready for writing.
+    fn epoll_mod(epoll_fd: RawFd, stream_fd: RawFd, evset: epoll::Events) -> Result<()> {
+        let event = epoll::Event::new(evset, stream_fd as u64);
+        epoll::ctl(
+            epoll_fd,
+            epoll::ControlOptions::EPOLL_CTL_MOD,
+            stream_fd,
+            event,
+        )
+        .map_err(ServerError::IOError)
+    }
+
+    /// Adds a stream to the `epoll` notification structure with the `EPOLLIN` event set.
+    fn epoll_add(epoll_fd: RawFd, stream_fd: RawFd) -> Result<()> {
+        epoll::ctl(
+            epoll_fd,
+            epoll::ControlOptions::EPOLL_CTL_ADD,
+            stream_fd,
+            epoll::Event::new(epoll::Events::EPOLLIN, stream_fd as u64),
+        )
+        .map_err(ServerError::IOError)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::fs;
+    use std::os::unix::net::UnixStream;
+
+    use common::Body;
+    use std::io::Read;
+    use std::io::Write;
+
+    #[test]
+    fn test_wait_one_connection() {
+        let path_to_socket = "/tmp/test_socket_http_server1.sock";
```

Comment:
Hmm, how about using some kind of random number for the path instead of incrementing one number? Right now we run all tests on one thread, so we shouldn't have a problem. But it can be that we decide to run them in parallel it can happen that we just add the same path to two tests.

Reply:
The incrementing number is what makes it work even multithreaded, as each test uses its own server with its own unique socket. I ran the tests in a multithreaded environment with `cargo test -- --test-threads=6` multiple times and it works as intended. There is no risk of two tests using the same socket.

Reply:
Clarification: there is nothing wrong with the tests as they are. They work on multi-threaded as well.

But: there is no risk **now**. My comment was future looking. It is very easy to copy-paste the first lines of an existing test to create another test. I assume this will happen when people will add more tests to this module. It is very easy to miss that each test has a different socket path when you're doing the copy-pasta. It happened in the past, that's why we don't use hard coded paths any more, and use `tempfile` instead (just an example). As it was more of a future looking comment, you can disregard it if you think that's too much work to do. It's also not super relevant, and more of a preference.

Reply:
I would prefer leaving it as it is for now.